### PR TITLE
Total rewrite

### PR DESCRIPTION
**Pull request by tobhe**
## Modifications
- replaced selfmade config parser with yaml parser
- replaced HDBC with persist library
- improved aeson protocol parser
- removed redundant modules
## Additions
- implemented full protocol
- added Testclient which supports the protocol in `test/TestCli.hs`
- project now uses [stack](https://github.com/commercialhaskell/stack)

### CODE REVIEW

File: openage-masterserver.cabal [hunk 1]
```diff
@@ -13,28 +13,78 @@ license:             AGPL-3
...
+copyright:           2016, openage team
 category:            Network, Game
 build-type:          Simple
 extra-source-files:  README.md
 cabal-version:       >=1.10
 
-executable openage-masterserver
-  ghc-options:         -O2 -Wall -dynamic -threaded -fno-warn-missing-signatures -fwarn-tabs
+library
+  ghc-options:         -O2 -Wall -Werror -dynamic -threaded -fwarn-tabs -fwarn-incomplete-uni-patterns -fwarn-incomplete-record-updates
```
Comment (TheJJ) [inline openage-masterserver.cabal]: For portability reasons, `-Werror` is a bad idea, compiler updates might render the project unusable that way.

File: src/Masterserver/Config.hs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  , database :: !Value        -- ^ Database connection info
+  } deriving Show
+
+-- | Derives aeson toJSON and fromJSON instances for Config.
+$(deriveJSON defaultOptions ''Config)
+
+-- | Get PostgresConf from yaml file.
+getPostgresConf :: IO PostgresConf
+getPostgresConf = do
+  Just yaml <- decodeFile "etc/openage/masterserver.yaml"
```
Comment (TheJJ) [inline src/Masterserver/Config.hs:34]: Can we add a command line option that specifies the config location? Which has a default value of `/etc/openage/masterserver.yaml` so the systemd service file works without setting the location? (`--config` or `-c`)

File: src/Masterserver/Protocol.hs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+-- | Game participant, players ingame settings
+data Participant = Participant
+  { parName :: !AuthPlayerName
+  , parCiv :: !Text
+  , parTeam :: !Int
+  , parReady :: !Bool
+  } deriving Show
+
+newParticipant :: AuthPlayerName -> Bool -> Participant
+newParticipant parName parReady = Participant{ parCiv="Britain"
```
Comment (TheJJ) [inline src/Masterserver/Protocol.hs]: just use "None" here i'd say :) The server might be used with a game that doesn't have the brits.

**Comment by TheJJ** [comment]
Awesome work! :dancers:
Now only our client has to actually use it.

**Comment by TheJJ** [comment]
Nice. Do you wanna add the config file path option now or after merging it?

**Comment by tobhe** [comment]
i'm not sure, i'm currently thinking about a better way to store the config data and to make it possible to specify config settings as command line options, which might take some time. If you want to you can merge it now and i'm just gonna make command line options in the next merge.

**Comment by TheJJ** [comment]
Jup. Thx :D

**Pull request merged**